bench/workgen/workgen.cxx

```diff
@@ -248,6 +281,231 @@ WorkloadRunner::start_table_idle_cycle(WT_CONNECTION *conn)
     }
     return 0;
 }
+
+/*
+ * Get the cumulative size of all the files under the given directory.
+ */
+static uint32_t
+get_dir_size_mb(std::string dir)
+{
+    uint64_t result = 0;
+    for (const auto &path : std::filesystem::recursive_directory_iterator(dir)) {
+        try {
+            result += std::filesystem::file_size(path);
+        } catch (std::filesystem::filesystem_error &e) {
+            /*
+             * A file might be dropped between listing the directory contents and getting the file
+             * sizes. Ignore such errors.
+             */
+            ASSERT(std::string(e.what()).find("No such file or directory") != std::string::npos);
+            continue;
+        }
+    }
+    return result / (WT_THOUSAND * WT_THOUSAND);
+}
+
+/*
+ * This function will create "dynamic_create_count" tables every "dynamic_create_period" seconds. It
+ * will also monitor the database size and stop creating tables once the size crosses
+ * "dynamic_create_target". It will restart table creation if the database size drops below
+ * "dynamic_create_trigger".
+ */
+int
+WorkloadRunner::start_dynamic_create(WT_CONNECTION *conn)
+{
+    WT_SESSION *session;
+    if (conn->open_session(conn, nullptr, nullptr, &session) != 0) {
+        THROW("Error opening a session.");
+    }
+
+    WT_DECL_RET;
+    ContextInternal *icontext = _workload->_context->_internal;
+    char uri[BUF_SIZE];
+    uint64_t total_creates = 0;
+    uint32_t db_size = get_dir_size_mb(_wt_home);
+
+    // Initially we start creating tables if the database size is less than the create target.
+    bool creating = db_size < _workload->options.dynamic_create_target;
+
+    while (!stopping) {
+        /*
+         * If we have been creating tables, continue until we reach the create target size. If we
+         * have not been creating tables, begin to do so if the database size falls below the create
+         * trigger.
+         */
+        db_size = get_dir_size_mb(_wt_home);
+        if (creating) {
+            creating = db_size < _workload->options.dynamic_create_target;
+            if (!creating) {
+                VERBOSE(*_workload,
+                  "Stopped creating new tables. db_size now "
+                    << db_size << " MB has reached the create target of "
+                    << _workload->options.dynamic_create_target << " MB.");
+            }
+        } else {
+            creating = db_size < _workload->options.dynamic_create_trigger;
+            if (creating) {
+                VERBOSE(*_workload,
+                  "Started creating new tables. db_size now "
+                    << db_size << " MB has reached the create trigger of "
+                    << _workload->options.dynamic_create_trigger << " MB.");
+            }
+        }
+        if (!creating) {
+            sleep(_workload->options.dynamic_create_period);
+            continue;
+        }
+
+        for (int creates = 0; creates < _workload->options.dynamic_create_count;
```

Comment:
I agree this can be cleaned a bit. I will suggest we let it be for now and restructure a bit when we switch to using random tables. As that change will require a bit of restructuring in both create and drop if there is a collision. I will keep your comment in mind for that change.

Reply:
I'm okay with that.